In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [2]:
TurkeyHC = pd.read_excel("ASHTURR5.xlsx")
AustriaHC = pd.read_excel("ASHAUTR5.xlsx")
GermanyHC = pd.read_excel("ASHDEUR5.xlsx")
EgyptHC = pd.read_excel("ASHEGYR5.xlsx")
FranceHC = pd.read_excel("ASHFRAR5.xlsx")
IranHC = pd.read_excel("ASHIRNR5.xlsx")
JordanHC = pd.read_excel("ASHJORR5.xlsx")
NetherlandsHC = pd.read_excel("ASHNLDR5.xlsx")
SwedenHC = pd.read_excel("ASHSWER5.xlsx")

HC is the home context data. This is a questionnaire filled in by the parents of each student. It contains crucial information on the immigration status of students and the languages they speak.

In [3]:
HCdf = pd.concat([TurkeyHC, AustriaHC, GermanyHC, EgyptHC, FranceHC, IranHC, JordanHC, NetherlandsHC, SwedenHC])

In [4]:
HCdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 5174
Columns: 118 entries, IDCNTRY to SCOPE
dtypes: float64(8), int64(6), object(104)
memory usage: 45.7+ MB


It is interesting here that when the dataframes were concatenated the Dtype of the columns changed from int to object. I believe this is because the data for some countries has been inputted as strings and for other countries as int. 

In [5]:
HCcolumns_to_keep = ['IDCNTRY','IDSTUD','ASBH02A','ASBH02B','ASBH03A','ASBH04', 'ASBH15A','ASBH15B','ASBH16','ASBH17A','ASBH17B','ASBH18AA','ASBH18AB']

In [7]:
HCdf = HCdf[HCcolumns_to_keep]

In [8]:
HCdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 5174
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   IDCNTRY   50367 non-null  int64 
 1   IDSTUD    50367 non-null  int64 
 2   ASBH02A   41503 non-null  object
 3   ASBH02B   16269 non-null  object
 4   ASBH03A   41179 non-null  object
 5   ASBH04    41231 non-null  object
 6   ASBH15A   37405 non-null  object
 7   ASBH15B   34597 non-null  object
 8   ASBH16    40406 non-null  object
 9   ASBH17A   36503 non-null  object
 10  ASBH17B   32361 non-null  object
 11  ASBH18AA  40161 non-null  object
 12  ASBH18AB  35131 non-null  object
dtypes: int64(2), object(11)
memory usage: 5.4+ MB


SC is the Student Context data which is a questionnaire filled in by the students themselves as well as some totals on their achievement in the test. 

In [9]:
TurkeySC = pd.read_excel("ASGTURR5.xlsx")
AustriaSC = pd.read_excel("ASGAUTR5.xlsx")
EgyptSC = pd.read_excel("ASGEGYR5.xlsx")
FranceSC = pd.read_excel("ASGFRAR5.xlsx")
IranSC = pd.read_excel("ASGIRNR5.xlsx")
JordanSC = pd.read_excel("ASGJORR5.xlsx")
NetherlandsSC = pd.read_excel("ASGNLDR5.xlsx")
SwedenSC = pd.read_excel("ASGSWER5.xlsx")
GermanySC = pd.read_excel("ASGDEUR5.xlsx")

In [10]:
SCdf = pd.concat([TurkeySC, AustriaSC, EgyptSC, FranceSC, IranSC, JordanSC, NetherlandsSC, SwedenSC, GermanySC])

In [11]:
column_list = SCdf.columns.to_list()
# Join the list into a single string separated by ', '
# Format each column name with quotes
formatted_columns = ', '.join(f"'{col}'" for col in column_list)

# Print the formatted string
print(formatted_columns)

'IDCNTRY', 'IDPOP', 'IDGRADER', 'IDGRADE', 'WAVE', 'IDSCHOOL', 'IDCLASS', 'IDSTUD', 'ITSEX', 'ITADMINI', 'LCID_SA', 'LCID_SQ', 'ITLANG_SA', 'ITLANG_SQ', 'IDBOOK', 'ASBG01', 'ASBG03', 'ASBG04', 'ASBG05A', 'ASBG05B', 'ASBG05C', 'ASBG05D', 'ASBG05E', 'ASBG05F', 'ASBG05G', 'ASBG05H', 'ASBG05I', 'ASBG05J', 'ASBG05K', 'ASBG06', 'ASBG07A', 'ASBG07B', 'ASBG08A', 'ASBG08B', 'ASBG09A', 'ASBG09B', 'ASBG09C', 'ASBG09D', 'ASBG09E', 'ASBG09F', 'ASBG09G', 'ASBG09H', 'ASBG10A', 'ASBG10B', 'ASBG10C', 'ASBG10D', 'ASBG10E', 'ASBG10F', 'ASBG11A', 'ASBG11B', 'ASBG11C', 'ASBG11D', 'ASBG11E', 'ASBG11F', 'ASBG11G', 'ASBG11H', 'ASBG11I', 'ASBG11J', 'ASBR01A', 'ASBR01B', 'ASBR01C', 'ASBR01D', 'ASBR01E', 'ASBR01F', 'ASBR01G', 'ASBR01H', 'ASBR01I', 'ASBR02A', 'ASBR02B', 'ASBR02C', 'ASBR02D', 'ASBR02E', 'ASBR03A', 'ASBR03B', 'ASBR03C', 'ASBR04', 'ASBR05', 'ASBR06A', 'ASBR06B', 'ASBR07A', 'ASBR07B', 'ASBR07C', 'ASBR07D', 'ASBR07E', 'ASBR07F', 'ASBR07G', 'ASBR07H', 'ASBR08A', 'ASBR08B', 'ASBR08C', 'ASBR08D', 'ASBR08

In [12]:
SCdemographic_info_columns = ['IDCNTRY','IDSTUD','ASBG01', 'ASBG03']

In [13]:
SCexperience_in_school_columns = ['ASBG10A', 'ASBG10B', 'ASBG10C', 'ASBG10D', 'ASBG10E', 'ASBG10F', 'ASBG11A', 'ASBG11B', 'ASBG11C', 'ASBG11D', 'ASBG11E', 'ASBG11F', 'ASBG11G', 'ASBG11H', 'ASBG11I', 'ASBG11J']

In [14]:
SCassessment_score_columns = ['ASRREA01', 'ASRREA02', 'ASRREA03', 'ASRREA04', 'ASRREA05', 'ASRLIT01', 'ASRLIT02', 'ASRLIT03', 'ASRLIT04', 'ASRLIT05', 'ASRINF01', 'ASRINF02', 'ASRINF03', 'ASRINF04', 'ASRINF05', 'ASRIIE01', 'ASRIIE02', 'ASRIIE03', 'ASRIIE04', 'ASRIIE05', 'ASRRSI01', 'ASRRSI02', 'ASRRSI03', 'ASRRSI04', 'ASRRSI05']

In [18]:
allSCcolumns = SCdemographic_info_columns + SCexperience_in_school_columns +SCassessment_score_columns

In [19]:
SCdf = SCdf[allSCcolumns]

In [20]:
SCdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 4610
Data columns (total 45 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   IDCNTRY   50367 non-null  int64  
 1   IDSTUD    50367 non-null  int64  
 2   ASBG01    49428 non-null  object 
 3   ASBG03    48168 non-null  object 
 4   ASBG10A   48311 non-null  object 
 5   ASBG10B   48077 non-null  object 
 6   ASBG10C   47893 non-null  object 
 7   ASBG10D   47822 non-null  object 
 8   ASBG10E   47844 non-null  object 
 9   ASBG10F   47945 non-null  object 
 10  ASBG11A   47967 non-null  object 
 11  ASBG11B   47756 non-null  object 
 12  ASBG11C   47634 non-null  object 
 13  ASBG11D   47497 non-null  object 
 14  ASBG11E   47543 non-null  object 
 15  ASBG11F   47843 non-null  object 
 16  ASBG11G   47707 non-null  object 
 17  ASBG11H   47701 non-null  object 
 18  ASBG11I   47647 non-null  object 
 19  ASBG11J   47613 non-null  object 
 20  ASRREA01  50367 non-null  flo

In [21]:
df = pd.merge(HCdf, SCdf, on=None)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 50366
Data columns (total 56 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   IDCNTRY   50367 non-null  int64  
 1   IDSTUD    50367 non-null  int64  
 2   ASBH02A   41503 non-null  object 
 3   ASBH02B   16269 non-null  object 
 4   ASBH03A   41179 non-null  object 
 5   ASBH04    41231 non-null  object 
 6   ASBH15A   37405 non-null  object 
 7   ASBH15B   34597 non-null  object 
 8   ASBH16    40406 non-null  object 
 9   ASBH17A   36503 non-null  object 
 10  ASBH17B   32361 non-null  object 
 11  ASBH18AA  40161 non-null  object 
 12  ASBH18AB  35131 non-null  object 
 13  ASBG01    49428 non-null  object 
 14  ASBG03    48168 non-null  object 
 15  ASBG10A   48311 non-null  object 
 16  ASBG10B   48077 non-null  object 
 17  ASBG10C   47893 non-null  object 
 18  ASBG10D   47822 non-null  object 
 19  ASBG10E   47844 non-null  object 
 20  ASBG10F   47945 non-null  ob

making new column - minage arrival - should i do this later?

In [ ]:
def populate_column(row):
    if row['ASBH02B'] == 1:
        return 0
    elif row['ASBH02B'] == 2:
        return 3
    elif row['ASBH02B'] == 3:
        return 6
    elif row['ASBH02B'] == 4:
        return 8
    else:
        return None  # Or any default value

# Apply the function to each row
DF['MINAGEARRIVAL'] = DF.apply(populate_column, axis=1)

In [ ]:
DF['MINAGEARRIVAL'].value_counts()

In [ ]:
DF['rounded_ages'] = DF['ASDAGE'].round()

In [ ]:
DF['rounded_ages'].value_counts()

First thing I need to do is recognise these countries and add these values to that column. 

In [23]:
df= df.rename(columns={'IDCNTRY': 'Country'})

# Mapping dictionary
update_dict = {40: 'Austria', 818: 'Egypt', 250:'France', 276:'Germany',364:'Iran',400:'Jordan',528:'Netherlands',752:'Sweden',792:'Turkey'}

# Updating the column using map
df['Country'] = df['Country'].map(update_dict).fillna(df['Country'])

In [24]:
df['Country'].value_counts()

Egypt          7979
Jordan         6150
Turkey         6032
Iran           5962
France         5339
Sweden         5175
Austria        4806
Germany        4611
Netherlands    4313
Name: Country, dtype: int64

## Now we are uniforming-ising the columns that are not all one type

maybe i should one hot encode these?

In [25]:
df['ASBH02A'].value_counts()

Yes    25072
1.0    13403
No      2342
2.0      548
9.0      138
Name: ASBH02A, dtype: int64

In [26]:
# Mapping dictionary
update_dict2 = {1.0: 'Yes',2.0:'No' }

# Updating the column using map
df['ASBH02A'] = df['ASBH02A'].map(update_dict2).fillna(df['ASBH02A'])
df['ASBH02A'].value_counts()

Yes    38475
No      2890
9.0      138
Name: ASBH02A, dtype: int64

So we have 2890 children who were not born in the country where they took the test.

In [27]:
df["ASBH03A"].value_counts()

Yes    25915
1.0    12270
2.0     1477
No      1175
9.0      342
Name: ASBH03A, dtype: int64

In [28]:
# Updating the column using map
df['ASBH03A'] = df['ASBH03A'].map(update_dict2).fillna(df['ASBH03A'])
df['ASBH03A'].value_counts()

Yes    38185
No      2652
9.0      342
Name: ASBH03A, dtype: int64

And we have 2652 children who did not speak the language of the test before they started school. So some children were born elsewhere but were already speaking the language of the test before they started school.

I want to see how many columns are affected by having strings and ints - so I want to see th unique values for each column. 

In [29]:
# Apply unique() to each column
unique_values = df.apply(lambda col: col.unique())

print(unique_values)

Country     [Turkey, Austria, Germany, Egypt, France, Iran...
IDSTUD      [50010601, 50010602, 50010603, 50010604, 50010...
ASBH02A                                   [Yes, nan, No, 9.0]
ASBH02B     [6.0, nan, 2.0, 1.0, 3.0, 9.0, 4.0, 3 to 5 yea...
ASBH03A                                   [Yes, No, 9.0, nan]
ASBH04      [1.0, 2.0, nan, 9.0, 4.0, 3.0, Always, Never, ...
ASBH15A     [99.0, 5.0, 2.0, 4.0, 7.0, 1.0, 6.0, 3.0, nan,...
ASBH15B     [7.0, 3.0, 99.0, 6.0, 4.0, 5.0, 8.0, 2.0, nan,...
ASBH16      [6.0, 3.0, 5.0, 4.0, nan, 9.0, 2.0, 1.0, Finis...
ASBH17A     [99.0, 1.0, 3.0, 4.0, 10.0, 12.0, 11.0, nan, 8...
ASBH17B     [9.0, 2.0, 99.0, 5.0, 3.0, 6.0, 12.0, 4.0, 10....
ASBH18AA                   [9.0, 1.0, 2.0, nan, 6.0, Yes, No]
ASBH18AB                   [1.0, 9.0, 2.0, nan, 6.0, Yes, No]
ASBG01                    [2.0, 1.0, nan, Girl, Boy, <Other>]
ASBG03      [1.0, 3.0, 2.0, 9.0, 4.0, nan, I always speak ...
ASBG10A     [1.0, 2.0, 3.0, 9.0, 4.0, nan, Agree a little,...
ASBG10B 

In [30]:
ordinal_mapping_frequency = {
    'At least once a week': 1,
    'Once or twice a month': 2,
    'A few times a year': 3,
    'Never': 4
}

In [32]:
experience_in_school_columns = ['ASBG11A',
 'ASBG11B',
 'ASBG11C',
 'ASBG11D',
 'ASBG11E',
 'ASBG11F',
 'ASBG11G',
 'ASBG11H',
 'ASBG11I',
 'ASBG11J']

In [33]:
for column in experience_in_school_columns:
    df[column]= df[column].map(ordinal_mapping_frequency)

In [ ]:
pd.set_option('display.max_colwidth', None)  # Optional: Set max column width
pd.set_option('display.expand_frame_repr', False)  # Optional: Do not wrap line

In [34]:
# Apply unique() to each column
unique_values_school_experience = df[experience_in_school_columns].apply(lambda col: col.unique())

print(unique_values_school_experience)

   ASBG11A  ASBG11B  ASBG11C  ASBG11D  ASBG11E  ASBG11F  ASBG11G  ASBG11H  \
0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
1      4.0      4.0      3.0      2.0      4.0      4.0      4.0      3.0   
2      2.0      2.0      1.0      4.0      3.0      3.0      2.0      4.0   
3      3.0      3.0      2.0      3.0      2.0      2.0      3.0      1.0   
4      1.0      1.0      4.0      1.0      1.0      1.0      1.0      2.0   

   ASBG11I  ASBG11J  
0      NaN      NaN  
1      4.0      4.0  
2      1.0      3.0  
3      3.0      1.0  
4      2.0      2.0  


In [35]:
ordinal_mapping_agreement = { 
    'Agree a lot': 1,
    'Agree a little': 2,
    'Disagree a little': 3,
    'Disagree a lot': 4
}

In [36]:
school_feeling_columns = ['ASBG10A', 'ASBG10B', 'ASBG10C', 'ASBG10D', 'ASBG10E', 'ASBG10F']

In [37]:
for column in school_feeling_columns:
    df[column]= df[column].map(ordinal_mapping_agreement)

In [38]:
# Apply unique() to each column
unique_values_school_feeling = df[school_feeling_columns].apply(lambda col: col.unique())

print(unique_values_school_feeling)

   ASBG10A  ASBG10B  ASBG10C  ASBG10D  ASBG10E  ASBG10F
0      NaN      NaN      NaN      NaN      NaN      NaN
1      2.0      3.0      2.0      2.0      3.0      1.0
2      1.0      2.0      1.0      1.0      1.0      2.0
3      3.0      1.0      3.0      3.0      2.0      3.0
4      4.0      4.0      4.0      4.0      4.0      4.0


In [39]:
ordinal_mapping_frequency = {
    'At least once a week': 1,
    'Once or twice a month': 2,
    'A few times a year': 3,
    'Never': 4
}

In [40]:
ordinal_mapping_education_levels = {
1: 'Did not go to school', 
 2: 'Some <Primary education—ISCED Level 1 or Lower secondary education—ISCED Level 2>', 
 3: '<Lower secondary education—ISCED Level 2>', 
 4: '<Upper secondary education—ISCED Level 3>', 
 5: '<Post-secondary, non-tertiary education— ISCED Level 4>', 
 6: '<Short-cycle tertiary education—ISCED Level 5>', 
 7: '<Bachelor’s or equivalent level—ISCED Level 6>', 
 8: '<Master’s or equivalent level—ISCED Level 7>', 
 9: '<Doctor or equivalent level—ISCED Level 8>',
 10: 'Not applicable'
 }

In [41]:
# Reverse the mapping dictionary
reverse_mapping_education_levels = {v: k for k, v in ordinal_mapping_education_levels.items()}

In [42]:
df['ASBH15A'] = df['ASBH15A'].map(reverse_mapping_education_levels)

In [43]:
df['ASBH15B'] = df['ASBH15B'].map(reverse_mapping_education_levels)

In [44]:
 ordinal_mapping_profession_levels = {
 1: 'Has never worked for pay',
 10: 'Professional Includes scientists,mathematicians; computer scientists,architects; engineers; life science and health pr',
 11: 'Technician or Associate Professional Includes science, engineering, and computer associates and technicians; life scienc',
 12: 'Not applicable',
 2: 'Small Business Owner Includes owners of small businesses (fewer than 25 employees) such as retail shops, services, resta',
 3: 'Clerical Worker Includes office clerks; secretaries; data entry operators; customer service clerks',
 4: 'Service or Sales Worker Includes travel attendants; restaurant service workers; personal care workers; protective servic',
 5: 'Skilled Agricultural or Fishery Worker Includes farmers; forestry workers; fishery workers',
 6: 'Craft or Trade Worker Includes builders, carpenters, plumbers, electricians, metal workers; machine mechanics; handicraf',
 7: 'Plant or Machine Operator Includes plant and machine operators; assembly-line operators; motor-vehicle drivers',
 8: 'General'
 }

In [45]:
# Reverse the mapping dictionary
reverse_mapping_profession_levels = {v: k for k, v in ordinal_mapping_profession_levels.items()}

In [46]:
df['ASBH17A'] = df['ASBH17A'].map(reverse_mapping_profession_levels)

In [47]:
df['ASBH17B'] = df['ASBH17B'].map(reverse_mapping_profession_levels)

In [48]:
ordinal_mapping_language = {
     1: 'I always speak <language of test> at home',
    2: 'I almost always speak <language of test> at home',
    3: 'I sometimes speak <language of test> and sometimes speak another language at home',
    4: 'I never speak <language of test> at home'
    
}

In [49]:
# Reverse the mapping dictionary
reverse_mapping_language = {v: k for k, v in ordinal_mapping_language.items()}
df['ASBG03'] = df['ASBG03'].map(reverse_mapping_language)
df['ASBG03'].unique()

array([nan,  1.,  3.,  2.,  4.])

In [50]:
ordinal_mapping_language_parents = {
    1: 'Always',
    2: 'Almost always',
    3: 'Sometimes',
    4: 'Never'
}

In [51]:
# Reverse the mapping dictionary
reverse_mapping_language_parents = {v: k for k, v in ordinal_mapping_language_parents.items()}
df['ASBH04'] = df['ASBH04'].map(reverse_mapping_language_parents)
df['ASBH04'].unique()

array([nan,  1.,  4.,  2.,  3.])

In [53]:
# Apply unique() to each column
unique_values = df.apply(lambda col: col.unique())

print(unique_values)

Country     [Turkey, Austria, Germany, Egypt, France, Iran...
IDSTUD      [50010601, 50010602, 50010603, 50010604, 50010...
ASBH02A                                   [Yes, nan, No, 9.0]
ASBH02B     [6.0, nan, 2.0, 1.0, 3.0, 9.0, 4.0, 3 to 5 yea...
ASBH03A                                   [Yes, No, 9.0, nan]
ASBH04                              [nan, 1.0, 4.0, 2.0, 3.0]
ASBH15A     [nan, 8.0, 4.0, 9.0, 6.0, 3.0, 2.0, 10.0, 7.0,...
ASBH15B     [nan, 8.0, 5.0, 9.0, 4.0, 6.0, 7.0, 3.0, 2.0, ...
ASBH16      [6.0, 3.0, 5.0, 4.0, nan, 9.0, 2.0, 1.0, Finis...
ASBH17A     [nan, 12.0, 11.0, 2.0, 4.0, 6.0, 7.0, 3.0, 5.0...
ASBH17B     [nan, 3.0, 11.0, 1.0, 4.0, 2.0, 6.0, 12.0, 5.0...
ASBH18AA                   [9.0, 1.0, 2.0, nan, 6.0, Yes, No]
ASBH18AB                   [1.0, 9.0, 2.0, nan, 6.0, Yes, No]
ASBG01                    [2.0, 1.0, nan, Girl, Boy, <Other>]
ASBG03                              [nan, 1.0, 3.0, 2.0, 4.0]
ASBG10A                             [nan, 2.0, 1.0, 3.0, 4.0]
ASBG10B 

In [54]:
# Define the function to map patterns to integers
def map_using_patterns(text):
    # Dictionary of regex patterns to integers
    pattern_to_integer = {
        r'Lower secondary education': 1,
        r'Upper secondary education': 2,
        r'Post-secondary, non-tertiary education': 3,
        r'Short-cycle tertiary education': 4,
        r'Bachelor’s or equivalent level': 5,
        r'Postgraduate degree': 6
    }

    # Check if the input is a string
    if isinstance(text, str):
        for pattern, integer in pattern_to_integer.items():
            if pd.Series([text]).str.contains(pattern).any():
                return integer
    return text  # Return the original value if it's not a string or no pattern matches

# Apply the function to the 'ASBH16' column
df['ASBH16'] = df['ASBH16'].apply(map_using_patterns)


In [55]:
df['ASBH16'].value_counts()

6.0    16285
5.0    11132
4.0     4655
2.0     4523
3.0     2452
1.0     1093
9.0      266
Name: ASBH16, dtype: int64

In [56]:
mapping_dict = {"Yes": 1, "No": 2}
df['ASBH18AA']= df['ASBH18AA'].map(mapping_dict)
df['ASBH18AB']= df['ASBH18AB'].map(mapping_dict)
df['ASBH18AA'].unique()
df['ASBH18AB'].unique()

array([nan,  1.,  2.])

In [57]:
mapping_dict_gender = {"Girl": 1, "Boy": 2, "<Other>": 3}
df['ASBG01']= df['ASBG01'].map(mapping_dict_gender)
df['ASBG01'].unique()

array([nan,  1.,  2.,  3.])

In [58]:
# Apply unique() to each column
unique_values_df = df.apply(lambda col: col.unique())

print(unique_values_df)

Country     [Turkey, Austria, Germany, Egypt, France, Iran...
IDSTUD      [50010601, 50010602, 50010603, 50010604, 50010...
ASBH02A                                   [Yes, nan, No, 9.0]
ASBH02B     [6.0, nan, 2.0, 1.0, 3.0, 9.0, 4.0, 3 to 5 yea...
ASBH03A                                   [Yes, No, 9.0, nan]
ASBH04                              [nan, 1.0, 4.0, 2.0, 3.0]
ASBH15A     [nan, 8.0, 4.0, 9.0, 6.0, 3.0, 2.0, 10.0, 7.0,...
ASBH15B     [nan, 8.0, 5.0, 9.0, 4.0, 6.0, 7.0, 3.0, 2.0, ...
ASBH16               [6.0, 3.0, 5.0, 4.0, nan, 9.0, 2.0, 1.0]
ASBH17A     [nan, 12.0, 11.0, 2.0, 4.0, 6.0, 7.0, 3.0, 5.0...
ASBH17B     [nan, 3.0, 11.0, 1.0, 4.0, 2.0, 6.0, 12.0, 5.0...
ASBH18AA                                      [nan, 1.0, 2.0]
ASBH18AB                                      [nan, 1.0, 2.0]
ASBG01                                   [nan, 1.0, 2.0, 3.0]
ASBG03                              [nan, 1.0, 3.0, 2.0, 4.0]
ASBG10A                             [nan, 2.0, 1.0, 3.0, 4.0]
ASBG10B 

In [60]:
df.to_excel('cleandata148.xlsx')

I want to remove those that have no answer to the asbh02a question because this is the most important for my analysis. 

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 50366
Data columns (total 56 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Country   50367 non-null  object 
 1   IDSTUD    50367 non-null  int64  
 2   ASBH02A   41503 non-null  object 
 3   ASBH02B   16269 non-null  object 
 4   ASBH03A   41179 non-null  object 
 5   ASBH04    27142 non-null  float64
 6   ASBH15A   23316 non-null  float64
 7   ASBH15B   20508 non-null  float64
 8   ASBH16    40406 non-null  float64
 9   ASBH17A   15016 non-null  float64
 10  ASBH17B   12722 non-null  float64
 11  ASBH18AA  26072 non-null  float64
 12  ASBH18AB  21042 non-null  float64
 13  ASBG01    37435 non-null  float64
 14  ASBG03    36188 non-null  float64
 15  ASBG10A   36332 non-null  float64
 16  ASBG10B   36097 non-null  float64
 17  ASBG10C   35913 non-null  float64
 18  ASBG10D   35842 non-null  float64
 19  ASBG10E   35864 non-null  float64
 20  ASBG10F   35965 non-null  fl

In [62]:
df = df.dropna(subset= ['ASBH02A'])

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41503 entries, 0 to 50364
Data columns (total 56 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Country   41503 non-null  object 
 1   IDSTUD    41503 non-null  int64  
 2   ASBH02A   41503 non-null  object 
 3   ASBH02B   16269 non-null  object 
 4   ASBH03A   40913 non-null  object 
 5   ASBH04    26888 non-null  float64
 6   ASBH15A   23005 non-null  float64
 7   ASBH15B   20263 non-null  float64
 8   ASBH16    40023 non-null  float64
 9   ASBH17A   14850 non-null  float64
 10  ASBH17B   12612 non-null  float64
 11  ASBH18AA  25767 non-null  float64
 12  ASBH18AB  20840 non-null  float64
 13  ASBG01    29560 non-null  float64
 14  ASBG03    28488 non-null  float64
 15  ASBG10A   28688 non-null  float64
 16  ASBG10B   28500 non-null  float64
 17  ASBG10C   28378 non-null  float64
 18  ASBG10D   28306 non-null  float64
 19  ASBG10E   28339 non-null  float64
 20  ASBG10F   28396 non-null  fl

And I want to remove the 9s in this column too. 

In [64]:
df = df[df['ASBH02A'] != 9]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41365 entries, 0 to 50364
Data columns (total 56 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Country   41365 non-null  object 
 1   IDSTUD    41365 non-null  int64  
 2   ASBH02A   41365 non-null  object 
 3   ASBH02B   16131 non-null  object 
 4   ASBH03A   40775 non-null  object 
 5   ASBH04    26888 non-null  float64
 6   ASBH15A   23005 non-null  float64
 7   ASBH15B   20263 non-null  float64
 8   ASBH16    39885 non-null  float64
 9   ASBH17A   14850 non-null  float64
 10  ASBH17B   12612 non-null  float64
 11  ASBH18AA  25767 non-null  float64
 12  ASBH18AB  20840 non-null  float64
 13  ASBG01    29510 non-null  float64
 14  ASBG03    28438 non-null  float64
 15  ASBG10A   28638 non-null  float64
 16  ASBG10B   28451 non-null  float64
 17  ASBG10C   28329 non-null  float64
 18  ASBG10D   28259 non-null  float64
 19  ASBG10E   28291 non-null  float64
 20  ASBG10F   28349 non-null  fl

In [65]:
df.to_excel('cleandata158.xlsx')